File: DesignPatterns/SingletonDesignPattern/CPP/logger.cpp
#include <iostream>
#include "logger.hpp"
using namespace std;

int Logger::ctr = 0;

Logger *Logger::loggerInstance = nullptr;

mutex Logger::mtx;

Logger::Logger() {
    ctr++;
    cout << "Logger instance created! No. of instances: " << ctr << endl;
}

void Logger::Log(string msg) {
    cout << msg << endl;
}

Logger *Logger::getLogger() {
    // double checked logging
    if (loggerInstance == nullptr) {
        mtx.lock();
        if (loggerInstance == nullptr) 
        {
            loggerInstance = new Logger();
        }
        mtx.unlock();
    }
    
    return loggerInstance;
}

File: DesignPatterns/SingletonDesignPattern/Python/logger.py
import threading

class Logger:
    _instance = None
    _lock = threading.Lock()
    ctr = 0

    def __new__(self):
        if self._instance == None:
            with self._lock:
                if self._instance == None:
                    self._instance = super(Logger, self).__new__(self)
                    self.ctr += 1
                    print(f"Logger instance created. No. of instances = {self.ctr}")
        return self._instance
    
    def log(self, msg):
        print(msg)
